Update DHO lifetime on entry lifetime change

ekrctb opened the pull request:
Previously, a `DrawableHitObject`'s lifetime can get out of sync with its `HitObjectLifetimeEntry`'s lifetime if the entry lifetime is modified.
The behavior could be observed by using the drawable visualizer to watch `LifetimeStart` and `LifetimeEnd` of a DHO in an editor. The lifetime changes weirdly when the hit object is selected and moved in the timeline. I'm not sure this behavior is noticeable without the draw visualizer.
Also, the behavior was confusing when debugging.

To reset lifetime when the `ScrollingHitObjectContainer` layout changes, it needs to set the lifetime of non-alive hit objects. Without this PR, the lifetime change will be partially ignored when DHO was alive.

This implementation is different from what #12878 has reverted, and is compatible with `LifetimeManagementContainer`.

ekrctb committed:
Update drawable lifetime on entry lifetime change

ekrctb committed:
Test drawable lifetime change on entry lifetime change

bdach commented on osu.Game/Rulesets/Objects/Pooling/PoolableDrawableWithLifetime.cs:
```diff
@@ -26,17 +27,18 @@ public abstract class PoolableDrawableWithLifetime<TEntry> : PoolableDrawable wh
-                base.LifetimeStart = value;
+                if (Entry == null && LifetimeStart != value)
+                    throw new InvalidOperationException($"Cannot modify lifetime of {nameof(PoolableDrawableWithLifetime<TEntry>)} when entry is not set");
 
-                if (Entry != null)
-                    Entry.LifetimeStart = value;
+                if (Entry == null) return;
```
The way this is written feels strange/backwards to me. Wouldn't it be better reading written as

```csharp
if (LifetimeEnd == value)
    return;

if (Entry == null)
    throw new InvalidOperationException($"Cannot modify lifetime of {nameof(PoolableDrawableWithLifetime<TEntry>)} when entry is not set");
```

ekrctb replied:
That changes the behavior because setting lifetime while `HitObjectLifetimeEntry.KeepAlive` is true changes "real lifetime". 
The behavior is tested in <https://github.com/ppy/osu/blob/b7afea37c4f8facce6a43e362371c2470110572c/osu.Game.Tests/Gameplay/TestSceneDrawableHitObject.cs#L43>.
I could add a comment, though.

ekrctb replied:
I think this behavior is confusing and I want it to be dropped. But I'll leave this as is to keep the scope of this PR minimal.

bdach commented on osu.Game/Rulesets/Objects/Pooling/PoolableDrawableWithLifetime.cs:
```diff
@@ -117,11 +121,23 @@ private void free()
+            setLifetimeFromEntry();
+        }
+
+        private void setLifetimeFromEntry()
+        {
+            base.LifetimeStart = Entry?.LifetimeStart ?? double.MinValue;
+            base.LifetimeEnd = Entry?.LifetimeEnd ?? double.MaxValue;
+        }
```
This is also a bit baffling to me - a `setLifetimeFromEntry()` method that is not only null-tolerant wrt the entry, but also _does things_ when there is no entry. I'd just inline the `{Min,Max}Value` restoration in `free()` if anything, and keep this code path as one that actually needs an entry to work.

bdach commented on osu.Game/Rulesets/Objects/Pooling/PoolableDrawableWithLifetime.cs:
```diff
@@ -26,17 +27,18 @@ public abstract class PoolableDrawableWithLifetime<TEntry> : PoolableDrawable wh
-                    Entry.LifetimeStart = value;
+                if (Entry == null) return;
+
+                Entry.LifetimeStart = value;
+                base.LifetimeStart = Entry.LifetimeStart;
```
Can use compound assignment:

```suggestion
                base.LifetimeStart = Entry.LifetimeStart = value;
```

Same goes for `End`, obviously.

ekrctb replied:
This transformation was incorrect when `KeepAlive` is true.

bdach replied:
How does that flag make my proposal incorrect? Because that sounds broken to hell. It's been cited twice now as a reason for writing really awkward code. And *if* it turns out that there is a *very* good reason for splitting the assignment, it better be commented.

ekrctb replied:
While `HitObjectLifetimeEntry.KeepAlive` is true, lifetime change is blocked and `Entry.LifetimeStart` will return `double.MinValue` rather than the value that is just assigned.
To sync the drawable lifetime and the entry lifetime, that value must be assigned to `base.LifetimeStart`.
If the compound assignment is used, it will be `base.LifetimeStart = value`.

bdach replied:
I'm confused. Assignment is right-associative, is it not? There should be no functional difference in execution.

That said I really don't like that the order matters here, but alas.

bdach replied:
Never mind, I get it. It still feels quite bad but can't see a way to avoid it.

ekrctb replied:
It turns out setting `base.LifetimeStart` is not required here because entry lifetime will cause lifetime change event and lifetime is synced in the callback.

bdach reviewed (commented):
Aside from some constructs I consider a bit weird I don't see huge issues with this. I checked that spinners work this time, at the least.

ekrctb committed:
Apply code styling suggestions

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

ekrctb committed:
Revert invalid code transformation

ekrctb committed:
Add test catching lifetime change while KeepAlive is true

ekrctb committed:
Fix wrong code

ekrctb committed:
Simplify lifetime setter

Setting entry lifetime will cause `LifetimeChanged` event and `base.LifetimeStart`/`End` will be modified in the callback.

smoogipoo reviewed (approved):
lgtm
